src/control/server/harness.go

```diff
@@ -228,6 +231,61 @@ func (h *IOServerHarness) startInstances(ctx context.Context, membership *system
 	return nil
 }
 
+// StopInstances will stop harness-managed instances.
+//
+// Iterate over instances and call Stop(signal) on each, return when all
+// instances are stopped or context is done.
+func (h *IOServerHarness) StopInstances(log logging.Logger, ctx context.Context, signal os.Signal, ranks ...uint32) error {
+	if !h.IsStarted() {
+		return FaultHarnessNotStarted
+	}
+
+	stopErrors := make([]error, 0, len(h.instances))
+	for _, i := range h.Instances() {
+		if !i.hasSuperblock() { // critical error
+			return FaultInstanceNoSuperblock(i.Index())
+		}
+
+		rank, ok := validateInstanceRank(log, i, ranks)
+		if !ok { // filtered out, no result expected
+			continue
+		}
+
+		if !i.IsStarted() {
+			continue
+		}
+
+		if err := i.Stop(signal); err != nil {
+			stopErrors = append(stopErrors, errors.Wrapf(err,
+				"rank %d sent signal %s", *rank, signal))
+		}
+	}
```

Comment:
Instead of stopping each instance sequentially and then polling for status with a hard-coded Sleep() in another goroutine below, I think it would be much more reliable to fan out to shut the instances down concurrently (each stop in a goroutine which feeds results back via an error channel) and then just poll using a for/select.

e.g.
```go
	instances := h.Instances()
	stopErrCh := make(chan error, len(instances))
	stopping := 0
	for _, i := range instances {
		if !i.hasSuperblock() { // critical error
			return FaultInstanceNoSuperblock(i.Index())
		}

		_, ok := validateInstanceRank(log, i, ranks)
		if !ok { // filtered out, no result expected
			continue
		}

		if !i.IsStarted() {
			continue
		}

		// supply the instance as a param to the goroutine to avoid shadowing
		go func(toStop *IOServerInstance) {
			select {
			case <-ctx.Done():
				log.Errorf("context canceled while stopping instance %d", toStop.Index())
			case stopErrCh <- toStop.Stop(signal):
				return
			}
		}(i)
		stopping++
	}

	stopErrors := make([]error, 0, len(instances))
	for {
		select {
		case <-ctx.Done():
			log.Error("graceful shutdown did not complete")
			return nil
		case stopErr := <-stopErrCh:
			if stopErr != nil {
				stopErrors = append(stopErrors, stopErr)
			}
			stopping--
			if stopping < 1 {
				if len(stopErrors) > 0 {
					return common.ConcatErrors(stopErrors, nil)
				}
				return nil
			}
		}
	}
```